### ISSUE TITLE
zfs storage errors on newer VM hypervisor/CPU

### ISSUE THREAD

**Issue by justinpryzby** [openzfs/zfs#14275]
### System information
Type | Version/Name
 --- | ---
Distribution Name	| Centos
Distribution Version	| 7
Kernel Version	| 5.4.225-1.el7.elrepo
Architecture	| x86_64
OpenZFS Version	| zfs-2.1.6-1

### Describe the problem you're observing

I noticed storage errors after a remote admin migrated a VM to a newer vmware hypervisor.  Efforts to avoid the issue have included:

 - do a cold boot rather than a live migration;
 - upgrade to more recent kernel from elrepo;
 - add new storage device and migrate data off the old storage.  This seemed to avoid the issue: ran scrub 6+ times with no errors; previously this would "get stuck".  But then the pool showed errors overnight;

I'm wondering if this could be a processor-specific problem?  Feel free to close this if there's no reason to suspect that.  This is related to #14267.

The old host is:
```
Model name:            Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush **dts** mmx fxsr sse sse2 ss ht syscall nx rdtscp lm constant_tsc arch_perfmon **pebs** **bts** nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm pti ssbd ibrs ibpb stibp tsc_adjust arat flush_l1d arch_capabilities
```

The new host is:
```
Model name:            Intel(R) Xeon(R) Gold 6240R CPU @ 2.40GHz
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm pti ssbd ibrs ibpb stibp tsc_adjust arat flush_l1d arch_capabilities
```

### Describe how to reproduce the problem

Unknown.

### Include any warning/errors/backtraces from the system logs
```
Dec  9 07:34:46 x kernel: sd 0:0:2:0: [sdb] tag#101 timing out command, waited 1080s
Dec  9 07:34:46 x kernel: sd 0:0:2:0: [sdb] tag#101 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_OK
Dec  9 07:34:46 x kernel: sd 0:0:2:0: [sdb] tag#101 CDB: Write(10) 2a 00 4d 76 56 36 00 00 95 00
Dec  9 07:34:46 x kernel: blk_update_request: I/O error, dev sdb, sector 1299600950 op 0x1:(WRITE) flags 0x700 phys_seg 15 prio class 0
```

**Comment by justinpryzby** [comment]
It's a continuing issue, but I suggest to close the ticket if there's no troubleshooting steps to follow.

**Comment by justinpryzby** [comment]
366 days later, we hit this issue again when attempting to deploy services to a new VM running on the same problematic hypervisor.  The new VM is running debian 12 / Linux 6.1.0-18-amd64 / zfs 2.2.2.

**Comment by justinpryzby** [comment]
We finally found a workaround for this - the errors are avoided if we attach the drive to a NVME controller rather than a SCSI controller.

A quick summary in case any of these are a useful hint.

- the issue only seemed to affects ZFS (?), and only over SCSI controller;
- only affects one customer, running under vmware; I was unable to find any customer-specific things going on, and the issue persisted even after disabling various management services;
- worked fine under on HP GEN8 hypervisor blade, and under a cisco UCSB-B200-M4 blade, but failed under HPE Synergy 480 Gen10;
- issue is avoided when run with nolapic, but that means we end up with only 1 CPU, which probably somehow avoids the issue;
- issue affects both centos7 (tried kernels 3.10 and 5.4), and debian 12 (kernels 6.1 and 6.6);
- issue affects "thick" drive;
- anything else to add here, ask now before I forget.

**Comment by justinpryzby** [comment]
Might as well close this even though nothing happened ?

**Comment by amotin** [comment]
@justinpryzby Would you be so kind to explain how do you see command timeouts on SCSI layer related to ZFS?  I suppose ZFS was on top of that SCSI layer inside VM, not on the hypervisor below it?

**Comment by justinpryzby** [comment]
Right, these VMs are typically running on our customer's hypervisor.
They give us access to the VM, but not the hypervisor nor the backing
storage.  Our customers don't know anything about the ZFS layer running
inside.

I ask them to build a VM with large storage on a separate block device
/dev/sdb (typically scsi), which I use as an LVM PV.  ZFS is created on
top of one or more LVs.  I know that configuration is considered to be
not ideal, but it's what we've decided to use.  In particular, that
allows keeping backups on a filesystem other than ZFS (not just a
separate FS/pool).

**Comment by amotin** [comment]
I still don't see how SCSI timeouts under VM is related to ZFS inside.
